File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/EquipmentService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.equipmentdto.creation.EquipmentCreationDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.equipmentdto.detail.EquipmentDetailDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.equipmentdto.overview.EquipmentStatusOverviewDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.equipmentdto.search.EquipmentSearchDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.equipmentdto.update.EquipmentUpdateDto;
import at.ac.tuwien.sepr.groupphase.backend.entity.enums.RentalStatus;
import at.ac.tuwien.sepr.groupphase.backend.entity.equipment.Equipment;
import at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException;

import java.time.LocalDate;
import java.util.List;

/**
 * Service interface for managing equipment-related operations.
 */
public interface EquipmentService {

    /**
     * Retrieves a list of equipment filtered by the specified type.
     *
     * @param type the type of equipment to filter by (e.g., "helmet", "ski", "snowboard")
     * @return a list of {@link EquipmentDetailDto} representing the equipment information for the specified type
     */
    List<EquipmentDetailDto> equipmentByType(String type);

    /**
     * Retrieves detailed information about a specific equipment item based on its unique identifier.
     *
     * @param id the unique identifier of the equipment to retrieve
     * @return an {@link EquipmentDetailDto} containing the detailed information of the specified equipment
     */
    EquipmentDetailDto equipmentById(Long id);

    /**
     * Retrieves detailed information about a specific equipment item based on its unique barcodeID.
     *
     * @param barcodeId the unique barcodeID of the equipment to retrieve
     * @return an {@link EquipmentDetailDto} containing detailed information of the specified equipment.
     */
    EquipmentDetailDto equipmentByBarcodeId(String barcodeId);

    /**
     * Creates a new equipment entry in the system based on the provided creation data.
     *
     * @param dto the data transfer object containing the information needed to create the equipment
     * @return the created {@link Equipment} entity
     */
    List<EquipmentDetailDto> createEquipment(EquipmentCreationDto dto);

    /**
     * Deletes an equipment entry from the system based on the specified type and ID.
     *
     * @param id the unique identifier of the equipment to delete
     */
    void deleteEquipment(Long id);

    /**
     * Partially updates an existing piece of equipment.
     * Only the non-null fields provided in the {@code updateDto} will be applied to the existing entity.
     *
     * @param id the unique identifier of the equipment to update
     * @param updateDto the data transfer object containing the new values
     * @return an {@link EquipmentDetailDto} representing the updated equipment
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if no equipment with the given ID exists in the database
     * @throws IllegalArgumentException if the type specified in the {@code updateDto} does not match the actual type of the existing equipment
     */
    EquipmentDetailDto updateEquipment(Long id, EquipmentUpdateDto updateDto);

    /**
     * Searches for equipment based on dynamic criteria.
     * All properties within the {@code searchDto} are optional. If a property is {@code null},
     * it will be ignored during the search process. If the entire DTO is {@code null} or empty,
     * all available equipment will be returned.
     *
     * @param searchDto the data transfer object containing the optional filter parameters
     * @return a list of {@link EquipmentDetailDto} matching the given criteria; an empty list if no matches are found
     */
    List<EquipmentDetailDto> searchEquipment(EquipmentSearchDto searchDto);

    /**
     * Updates the RentalStatus of all the equipments in the provided list of EquipmentIds. The only allowed
     * RentalStatus-Changes are as follows:
     * AVAILABLE -> RENTED || MAINTENANCE
     * RENTED -> AVAILABLE
     * MAINTENANCE -> AVAILABLE
     *
     * @param equipmentIds a list of EquipmentIds for equipments' whose RentalStatus should be updated.
     * @param newRentalStatus the new RentalStatus the equipment's should be updated to
     */
    public void updateEquipmentStatuses(List<Long> equipmentIds, RentalStatus newRentalStatus);

    /**
     * Retrieves an aggregated overview of all equipment items, grouped by equipment type
     * and rental status.
     *
     * @return an {@link EquipmentStatusOverviewDto} containing counts per type and status,
     *         including zero-counts for combinations with no matching equipment
     */
    EquipmentStatusOverviewDto getStatusOverview();

}

File: frontend/src/app/services/equipment.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Globals} from '../global/globals';
import {Equipment} from '../dtos/equipment';
import {EquipmentCreation} from '../dtos/equipment-creation';
import {EquipmentUpdate} from '../dtos/equipment-update';
import {EquipmentSearch} from '../dtos/equipment-search';
import {EquipmentOverview} from "../dtos/equipment-overview";

@Injectable({
  providedIn: 'root'
})
export class EquipmentService {

  private equipmentBaseUri: string = this.globals.backendUri + '/equipment';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Loads all equipment from the backend.
   *
   * @return an Observable of all equipment
   */
  getAll(): Observable<Equipment[]> {
    return this.httpClient.get<Equipment[]>(this.equipmentBaseUri);
  }

  /**
   * Loads an aggregated overview of equipment counts, grouped by equipment type and rental status.
   *
   * @return an Observable of the equipment status overview
   */
  getStatusOverview(): Observable<EquipmentOverview> {
    return this.httpClient.get<EquipmentOverview>(`${this.equipmentBaseUri}/overview`);
  }

  /**
   * Loads equipment by id from the backend.
   *
   * @param id the unique id of the equipment
   * @return an Observable of the found Equipment
   */
  getById(id: number): Observable<Equipment> {
    return this.httpClient.get<Equipment>(`${this.equipmentBaseUri}/${id}`);
  }

  /**
   * Loads equipment by barcode ID from the backend.
   *
   * @param barcodeId the unique barcode string of the equipment
   * @return an Observable of the found Equipment
   */
  getByBarcodeId(barcodeId: string): Observable<Equipment> {
    return this.httpClient.get<Equipment>(`${this.equipmentBaseUri}/barcode/${barcodeId}`);
  }

  /**
   * Deletes equipment by id in backend.
   *
   * @param id the unique id of the equipment
   * @return an Observable of the deleted Equipment
   */
  delete(id: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.equipmentBaseUri}/${id}`);
  }

  /**
   * Creates new equipment in backend.
   *
   * @return an Observable of the newly created Equipment
   */
  create(equipment: EquipmentCreation): Observable<Equipment> {
    return this.httpClient.post<Equipment>(this.equipmentBaseUri, equipment);
  }

  /**
   * Updates equipment in backend.
   *
   * @return an Observable of the updated equipment
   */
  update(id: number, equipment: EquipmentUpdate): Observable<Equipment> {
    return this.httpClient.patch<Equipment>(`${this.equipmentBaseUri}/${id}`, equipment);

  }

  /**
   * Searches for pieces of equipment in backend as specified in the search parameters.
   *
   * @param search the searchParameters to be used in the search
   * @return an Observable of the found equipment that conforms to the given searchParameters
   */
  search(search: EquipmentSearch): Observable<Equipment[]> {
    let params = new HttpParams();
    if (search.model) {
      params = params.set('model', search.model);
    }
    if (search.type) {
      params = params.set('type', search.type);
    }
    if (search.status) {
      params = params.set('status', search.status);
    }
    if (search.targetSkillLevel) {
      params = params.set('targetSkillLevel', search.targetSkillLevel);
    }
    if (search.start) {
      params = params.set('start', search.start);
    }
    if(search.end) {
      params = params.set('end', search.end);
    }

    return this.httpClient.get<Equipment[]>(this.equipmentBaseUri, {params});
  }

}
